geode-core/src/main/java/org/apache/geode/internal/security/shiro/JMXShiroAuthenticator.java

```diff
@@ -76,8 +79,17 @@ public void handleNotification(Notification notification, Object handback) {
     if (notification instanceof JMXConnectionNotification) {
       JMXConnectionNotification cxNotification = (JMXConnectionNotification) notification;
       String type = cxNotification.getType();
-      if (JMXConnectionNotification.CLOSED.equals(type)) {
-        this.securityService.logout();
+      String connectionId = cxNotification.getConnectionId();
+      synchronized (connectedUsers) {
+        if (JMXConnectionNotification.OPENED.equals(type)) {
+          if (!connectedUsers.containsKey(connectionId)) {
+            connectedUsers.put(connectionId, securityService.getSubject());
```

Comment:
instead of doing both containsKey and put how about just doing putIfAbsent?

Reply:
thanks! this is the code copied from the original PR. Coming to think of it, I think it's better just to do `put` here. First of all, i can't think of any cases that a connection id would be used more than once in a OPEN event; secondly, even if it was, the last subject should win.